internal/tools/aws/database_multitenant.go

```diff
@@ -0,0 +1,730 @@
+		}
+	}
+
+	return unlockFN, nil
+}
+
+func (d *RDSMultitenantDatabase) validateMultitenantDatabaseInstallations(installations model.MultitenantDatabaseInstallationIDs, multitenantDatabase *model.MultitenantDatabase) error {
+	expectedInstallations, err := multitenantDatabase.GetInstallationIDs()
+	if err != nil {
+		return errors.Errorf("failed to get installations from multitenant database ID %s", multitenantDatabase.ID)
+	}
+
+	if len(installations) != len(expectedInstallations) {
+		return errors.Errorf("supplied %d installations, but multitenant database ID %s has %d", len(installations), multitenantDatabase.ID, len(expectedInstallations))
+	}
+
+	for _, installation := range installations {
+		if !expectedInstallations.Contains(installation) {
+			return errors.Errorf("unable to find installation ID %s in the multitenant database ID %s", installation, multitenantDatabase.ID)
+		}
+	}
+
+	return nil
+}
+
+func (d *RDSMultitenantDatabase) removeRDSDatabase(rdsDatabaseCluster *model.MultitenantDatabase, rdsDatabaseName string, store model.InstallationDatabaseStoreInterface, logger log.FieldLogger) error {
+	logger = logger.WithField("rds-cluster-id", rdsDatabaseCluster.ID)
+
+	unlocked, err := d.lockMultitenantDatabase(rdsDatabaseCluster.ID, store)
+	if err != nil {
+		return errors.Wrapf(err, "failed to lock multitenant RDS cluster ID %s", rdsDatabaseCluster.ID)
+	}
+	defer unlocked(logger)
+
+	rdsCluster, err := d.describeRDSCluster(rdsDatabaseCluster.ID)
+	if err != nil {
+		return errors.Wrapf(err, "failed to describe multitenant RDS cluster ID %s", rdsDatabaseCluster.ID)
+	}
+
+	masterSecretValue, err := d.client.Service().secretsManager.GetSecretValue(&secretsmanager.GetSecretValueInput{
+		SecretId: &rdsDatabaseCluster.ID,
+	})
+	if err != nil {
+		return errors.Wrapf(err, "failed to get master secret by ID %s", rdsDatabaseCluster.ID)
+	}
+
+	close, err := d.connectRDSCluster(rdsMySQLSchemaInformationDatabase, *rdsCluster.Endpoint, DefaultMattermostDatabaseUsername, *masterSecretValue.SecretString)
+	if err != nil {
+		return errors.Wrapf(err, "unable to connect to multitenant RDS cluster ID %s", rdsDatabaseCluster.ID)
+	}
+	defer close(logger)
+
+	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(DefaultMySQLContextTimeSeconds*time.Second))
+	defer cancel()
+
+	err = d.dropDatabaseIfExists(ctx, rdsDatabaseName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to drop multitenant RDS database name %s", rdsDatabaseName)
+	}
+
+	databaseInstallationIDs, err := store.RemoveMultitenantDatabaseInstallationID(rdsDatabaseCluster.ID, d.installationID)
+	if err != nil {
+		return errors.Wrapf(err, "failed to remove installation ID %s from datasore", d.installationID)
+	}
+
+	err = d.updateCounterTag(rdsCluster.DBClusterArn, len(databaseInstallationIDs))
+	if err != nil {
+		return errors.Wrapf(err, "failed to update counter tag for the multitenant RDS cluster ID %s", rdsDatabaseCluster.ID)
+	}
```

Comment:
This happens after the MultitenantStore has been updated to remove the installation ID. Due to this, there is the possibility to not update this tag. It's not a huge deal in this case, but the store should be the final thing to be updated.

Reply:
The code above always return the installation IDs for the given multitenant database. This is independent of the installation being removed or not. However, if the installation ID exists and this operation is called without an error being returned, it's guaranteed that the ID was removed. This simplifies the things in case the `updateCounterTag()` returns error and needs to retry.

Reply:
Hmm, I am not sure that I understand.

I am pointing out that if line 611 succeeds, but line 616 fails, then we will never reach this `removeRDSDatabase` method again with the current code and the `updateCounterTag` will never be successfully called.

Reply:
Yes, that's a fact. I didn't want move all this logic into a helper for this reason. Let me work on this and add testing around it.